FreeBSD system roots implementation

tdbhacks opened the pull request:
This PR is based off [#16083](https://github.com/grpc/grpc/pull/16083), with the implementation being branched from the existing linux-specific logic.

<!--

Your pull request will be routed to the following person by default for triaging.
If you know who should review your pull request, please remove the mentioning below.

-->

@drfloob

tdbhacks commented:
Note to reviewers: the related test now runs on both linux and freebsd, but is still named `linux_system_roots_test`. Happy to rename it if that's appropriate, naming suggestions are very welcome!

drfloob commented:
Thanks for the contribution, @tdbhacks. It's a bit tricky to evaluate since we don't have any test infrastructure for FreeBSD - the platform is entirely community-supported, and we don't have a FreeBSD-community point of contact for gRPC. We're generally open to accepting community patches like this, with the understanding they may break without continued community support (and we won't know when they do break).

This looks useful at first glance, but nearly identical to the linux version (only the paths differ, yes?). I think it would be better to unify the two.

@ZhenLian could you take a look at this as well? Thanks.

tdbhacks commented:
@drfloob thanks for taking a look!

Indeed, only the paths differ (other than the logic to check directories in `LoadSystemRootCerts()` being dropped, since we only use files here). Wasn't sure if this was worth an independent implementation, I agree that it's essentially duplicate code so I'm open to merging with the linux implementation. We could either just add the FreeBSD paths to the list of linux paths, or we could choose the right set of paths based on platform macros. The latter seems a little less clean, but I don't feel strongly so either option works for me :)

Side note: I'd like to do this for MacOS too, which at first glance seems to use `/etc/ssl/cert.pem` too, either in a different PR or here. Wdyt?

drfloob commented:
> We could either just add the FreeBSD paths to the list of linux paths, or we could choose the right set of paths based on platform macros.

For security reasons, I think we'd want to be explicit about using the appropriate canonical paths for specific platforms.

Given that Mac is a supported platform in gRPC, I'd recommend a separate PR. There are separate concerns there, and potentially reasons to reject or roll back commits that don't exist for FreeBSD.

tdbhacks commented:
Sending a separate PR for Mac SGTM.

In the meantime, I've merged the FreeBSD implementation as suggested, it's now choosing the right set of paths based on the platform.

ZhenLian commented on src/core/lib/security/security_connector/load_system_roots_linux.cc:
```diff
@@ -51,20 +51,27 @@ GPR_GLOBAL_CONFIG_DEFINE_STRING(grpc_system_ssl_roots_dir, "",
     "/etc/ssl/certs", "/system/etc/security/cacerts", "/usr/local/share/certs",
     "/etc/pki/tls/certs", "/etc/openssl/certs"};
+#endif // GPR_LINUX || GPR_ANDROID
+#ifdef GPR_FREEBSD
```
What's the difference between using `ifdef GPR_FREEBSD` and `if defined(GPR_FREEBSD)`? 
We probably want to keep the syntax consistent.

tdbhacks replied:
Thanks for taking a look! So `ifdef VAL` and `if defined(VAL)` are equivalent, the reason why `if defined(VAL)` is used above is that it allows multiple conditions to be combined into a single if statement (i.e. `ifdef GPR_LINUX || GPR_FREEBSD` would not work). See https://stackoverflow.com/questions/39290019/is-if-defined-macro-equivalent-to-ifdef-macro

Let me know if you'd still like me to change it, happy to do it if you think it's cleaner that way.

re: Mac PR: I haven't sent it yet because I'm wondering whether we should just rename this file, remove the "_linux" suffix, and add the Mac-specific path in here too, for the same reason why @drfloob suggested merging the FreeBSD and Linux implementations. That way we would only have this implementation for all platforms where we support loading system roots, and the fallback implementation for all other platforms. WDYT? If that sounds good to you I'll send in a follow-up PR once this is merged (unless you think it's fine to add both here).

ZhenLian replied:
I see, thanks for the explanation! Would you mind change this to `if defined(VAL)` to make it consistent?

ZhenLian reviewed (changes requested):
Thanks for making the contributions and sorry about the delay! I was just back from vacation.

The change overall looks good to me, except one minor thing. Let me open the pre-submit tests.
BTW, where is the change related to Mac system? Do you plan to work on it after this is merged?

ZhenLian commented:
Let me re-run the test. In the meantime, you could run `clang-format -i THE_CHANGED_FILE` to see anything is changed.

tdbhacks commented:
Thanks, re-ran clang-format and it appears to have done the trick, not sure why it didn't work last time. I've also changed `ifdef` to `if defined` as suggested, and squashed all my commits.

I'll send a follow-up PR once this is merged to do the same for MacOS. I believe @bdhess spoke with @veblush a few weeks back and he was okay with adding a MacOS implementation based on the same logic and using `/etc/ssl/cert.pem`.

markdroth commented on src/core/lib/security/security_connector/load_system_roots_linux.cc:
```diff
@@ -51,20 +51,27 @@ GPR_GLOBAL_CONFIG_DEFINE_STRING(grpc_system_ssl_roots_dir, "",
+const char* kCertDirectories[] = {
     "/etc/ssl/certs", "/system/etc/security/cacerts", "/usr/local/share/certs",
     "/etc/pki/tls/certs", "/etc/openssl/certs"};
+#endif  // GPR_LINUX || GPR_ANDROID
```
Instead of `#endif` and then another independent `#if`, please use `#elif`.  That way, we know for sure that only one of these two definitions will be used.

tdbhacks replied:
Good point, changed to `#elif`.

markdroth reviewed (commented):
Just one minor comment, otherwise this looks great!

markdroth commented:
Test failures look unrelated, so I'll go ahead and merge this.
